File: Algorithms/src/719. Find K-th Smallest Pair Distance/Code/solution.py
class Solution:
    def countPairs(self, nums, mid):
        """
        This function counts how many pairs (i, j) exist in the sorted array `nums` 
        such that the difference between nums[j] and nums[i] is less than or equal to `mid`.
        """
        count = 0  # Initialize the count of pairs
        j = 0      # Initialize the pointer j which will be used to find the upper bound of the pair

        # Iterate through each element in the array
        for i in range(len(nums)):
            # Move the pointer j to find the maximum index where the difference is <= mid
            while j < len(nums) and nums[j] - nums[i] <= mid:
                j += 1
            # The number of valid pairs with i as the first element is j - i - 1
            count += j - i - 1

        return count  # Return the total count of valid pairs

    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        """
        This function finds the k-th smallest distance pair (difference between any two elements)
        in the array `nums`.
        """
        nums.sort()  # Sort the array to enable binary search on pair differences
        
        # Initialize the binary search range for the distance between pairs
        low, high = 0, nums[-1] - nums[0]

        # Perform binary search
        while low < high:
            mid = (low + high) // 2  # Calculate the mid-point of the current range

            # Count the number of pairs with difference <= mid
            if self.countPairs(nums, mid) >= k:
                # If there are at least k pairs with difference <= mid, 
                # the answer lies in the lower half including mid
                high = mid
            else:
                # Otherwise, the answer lies in the upper half excluding mid
                low = mid + 1

        return low  # The k-th smallest distance is found

File: Algorithms/src/719. Find K-th Smallest Pair Distance/Code/solution.js
/**
 * @param {number[]} nums - An array of integers.
 * @param {number} k - The k-th smallest distance pair to find.
 * @return {number} - The k-th smallest distance pair.
 */
var smallestDistancePair = function (nums, k) {
  /**
   * Helper function to count the number of pairs with a distance
   * less than or equal to the given 'mid' value.
   *
   * @param {number[]} nums - Sorted array of integers.
   * @param {number} mid - The maximum allowed distance for pairs.
   * @return {number} - The count of pairs with a distance <= 'mid'.
   */
  const countPairs = (nums, mid) => {
    let count = 0; // Initialize the pair count to zero
    let j = 0; // Initialize the right pointer

    // Iterate through the array with the left pointer 'i'
    for (let i = 0; i < nums.length; i++) {
      // Move the right pointer 'j' to find pairs where the difference is <= 'mid'
      while (j < nums.length && nums[j] - nums[i] <= mid) {
        j++;
      }
      // The number of valid pairs with 'i' as the left element is (j - i - 1)
      count += j - i - 1;
    }
    return count; // Return the total count of valid pairs
  };

  // Sort the array to facilitate the binary search approach
  nums.sort((a, b) => a - b);

  // Define the search space for binary search
  let low = 0; // The minimum possible distance
  let high = nums[nums.length - 1] - nums[0]; // The maximum possible distance

  // Perform binary search to find the k-th smallest distance pair
  while (low < high) {
    // Calculate the midpoint of the current search space
    let mid = Math.floor((low + high) / 2);

    // Use the helper function to count pairs with distance <= 'mid'
    if (countPairs(nums, mid) >= k) {
      // If there are at least 'k' pairs with distance <= 'mid', search the lower half
      high = mid;
    } else {
      // Otherwise, search the upper half
      low = mid + 1;
    }
  }

  // The 'low' value will be the k-th smallest distance pair after the loop
  return low;
};
